File: programmers/level 1/행렬_테두리_회전하기.py
# 240404
def minus_one(x):
    return x - 1


def solution(rows, columns, queries):
    answer = []
    arr = [[0] * columns for _ in range(rows)]
    for i in range(rows * columns):
        arr[i // columns][i % columns] = i + 1

    for query in queries:
        y1, x1, y2, x2 = map(minus_one, query)

        temp = []
        for j in range(x1, x2):
            temp.append(arr[y1][j])

        for i in range(y1, y2):
            temp.append(arr[i][x2])

        for j in range(x2, x1, -1):
            temp.append(arr[y2][j])

        for i in range(y2, y1, -1):
            temp.append(arr[i][x1])

        # 변경
        idx = 0
        for j in range(x1 + 1, x2):
            arr[y1][j] = temp[idx]
            idx += 1

        for i in range(y1, y2):
            arr[i][x2] = temp[idx]
            idx += 1

        for j in range(x2, x1, -1):
            arr[y2][j] = temp[idx]
            idx += 1

        for i in range(y2, y1, -1):
            arr[i][x1] = temp[idx]
            idx += 1

        arr[y1][x1] = temp[-1]
        answer.append(min(temp))

    return answer

File: programmers/level 2/행렬_테두리_회전하기.js
// 250805
function solution(rows, columns, queries) {
    const answer = []
    const matrix = Array.from({length : rows}, (_, i) => Array.from({length : columns}, (_, j) => i*columns + j + 1))

    function elementsOf(r1, c1, r2, c2) {
        const tempEl = []
        const tempPos = []

        for (let j = c1; j < c2; j += 1) {
            tempEl.push(matrix[r1][j])
            tempPos.push([r1, j])
        }

        for (let i = r1; i < r2; i += 1) {
            tempEl.push(matrix[i][c2])
            tempPos.push([i, c2])
        }

        for (let j = c2; j > c1; j -= 1) {
            tempEl.push(matrix[r2][j])
            tempPos.push([r2, j])
        }

        for (let i = r2; i > r1; i -= 1) {
            tempEl.push(matrix[i][c1])
            tempPos.push([i, c1])
        }

        return [tempEl, tempPos]
    }

    function rotate(arr, position) {
        const n = arr.length
        for (let i = 0; i < n-1; i += 1) {
            const [r, c] = position[i+1]
            matrix[r][c] = arr[i]
        }
        const [r, c] = position[0]
        matrix[r][c] = arr.at(-1)
    }

    queries.forEach((query) => {
        const [arr, position] = elementsOf(...query.map((num) => num-1))
        answer.push(Math.min(...arr))
        rotate(arr, position)
    })

    return answer
}